src/System.Management.Automation/engine/lang/parserutils.cs

```diff
@@ -785,6 +785,95 @@ internal static object JoinOperator(ExecutionContext context, IScriptExtent erro
+        /// <param name="lval">The object on which to start.</param>
+        /// <param name="rval">The object on which to stop.</param>
+        /// <returns>The array of objects.</returns>
+        internal static object RangeOperator(ExecutionContext context, IScriptExtent errorPosition, object lval, object rval)
+        {
+            var lbase = PSObject.Base(lval);
+            var rbase = PSObject.Base(rval);
+
+            // If both arguments is [char] type or [string] type with length==1
+            // return objects of [char] type.
+            // In special case "0".."9" return objects of [int] type.
+            if (AsChar(lbase) is char lc && AsChar(rbase) is char rc)
+            {
+                return CharOps.Range(lc, rc);
+            }
+
+            // As a last resort, the range operator tries to return objects of [int] type.
+            //    1..10
+            //    "1".."10"
+            //    [int]"1"..[int]"10"
+            var l = Convert.ToInt32(lbase);
+            var r = Convert.ToInt32(rbase);
+
+            return IntOps.Range(l, r);
+        }
+
+        /// <summary>
+        /// The implementation of an enumerator for the PowerShell range operator.
+        /// </summary>
+        /// <param name="lval">The object on which to start.</param>
+        /// <param name="rval">The object on which to stop.</param>
+        /// <returns>The enumerator.</returns>
+        internal static IEnumerator GetRangeEnumerator(object lval, object rval)
+        {
+            var lbase = PSObject.Base(lval);
+            var rbase = PSObject.Base(rval);
+
+            // If both arguments is [char] type or [string] type with length==1
+            // return objects of [char] type.
+            // In special case "0".."9" return objects of [int] type.
+            if (AsChar(lbase) is char lc && AsChar(rbase) is char rc)
+            {
+                return new CharRangeEnumerator(lc, rc);
+            }
+
+            // As a last resort, the range operator tries to return objects of [int] type.
+            //    1..10
+            //    "1".."10"
+            //    [int]"1"..[int]"10"
+            var l = Convert.ToInt32(lbase);
+            var r = Convert.ToInt32(rbase);
+
+            return new RangeEnumerator(l, r);
+        }
+
+        // Help function for Range operator.
+        //
+        // In common case:
+        //      for [char] type
+        //      for [string] type and length == 1
+        // return objects of [char] type:
+        //     [char]'A'..[char]'Z'
+        //     [char]'A'..'Z'
+        //     [char]'A'.."Z"
+        //     'A'..[char]'Z'
+        //     "A"..[char]'Z'
+        //     "A".."Z"
+        //     [char]"A"..[string]"Z"
+        //     "A"..[char]"Z"
+        //     [string]"A".."Z"
+        // and so on.
+        //
+        // In special case:
+        //     [char]'0'..[char]'9'
+        //     "0".."9"
+        // return objects of [int] type.
```

Comment:
This comment only applies to the later case, not the former, right?

I don't recall the discussion around this, but I have a feeling we'll regret going this way despite it being a breaking change.

Reply:
Yes, I removed ` [char]'0'..[char]'9'`